File: 2024/day-2/part_1.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(void) {
  FILE *file = fopen("input", "r");

  if (!file) {
    perror("File error");

    return EXIT_FAILURE;
  }

  char string[50];
  int array[20];
  int safeCount = 0;

  while (fgets(string, 50, file)) {
    char *token = strtok(string, " ");
    int index = 0;

    while (token) {
      array[index++] = atoi(token);
      token = strtok(NULL, " ");
    }

    bool isIncreasing = true, isDecreasing = true, isValid = true;

    for (int i = 0; i < index - 1; ++i) {
      int diff = array[i] - array[i + 1];

      if (!diff || diff < -3 || diff > 3) {
        isValid = false;

        break;
      }

      if (diff < 0)
        isDecreasing = false;
      else if (diff > 0)
        isIncreasing = false;

      if (!isDecreasing && !isIncreasing) {
        isValid = false;

        break;
      }
    }

    if (isValid)
      ++safeCount;
  }

  printf("%d\n", safeCount);
  fclose(file);

  return EXIT_SUCCESS;
}

File: 2024/day-2/part_1.py
def main() -> None:
    safe_count: int = 0

    with open("input", "r", encoding="UTF-8") as file:
        for line in file:
            array: list[int] = list(map(int, line.strip().split()))

            is_increasing: bool
            is_decreasing: bool
            is_valid: bool

            is_increasing = is_decreasing = is_valid = True

            for _ in range(len(array) - 1):
                diff = array[_] - array[_ + 1]

                if not diff or diff < -3 or diff > 3:
                    is_valid = False

                    break

                if diff < 0:
                    is_decreasing = False
                elif diff > 0:
                    is_increasing = False

                if not is_decreasing and not is_increasing:
                    is_valid = False

                    break

            if is_valid:
                safe_count += 1

    print(safe_count)


if __name__ == "__main__":
    main()
